loopy/transform/loop_fusion.py

```diff
@@ -0,0 +1,935 @@
+        )
+        not_seen_inames = frozenset(
+            iname
+            for iname in iname_chain
+            if iname not in self.iname_to_tree_node_id
+        )
+        seen_inames = iname_chain - not_seen_inames
+
+        # {{{ verbatim copied from loopy/schedule/tools.py
+
+        from loopy.schedule.tools import _add_inner_loops, separate_loop_nest
+
+        all_nests = {self.iname_to_tree_node_id[iname] for iname in seen_inames}
+
+        self.tree, outer_loop, inner_loop = separate_loop_nest(
+            self.tree, (all_nests | {frozenset()}), seen_inames
+        )
+        if not_seen_inames:
+            # make '_not_seen_inames' nest inside the seen ones.
+            # example: if there is already a loop nesting "i,j,k"
+            # and the current iname chain is "i,j,l". Only way this is possible
+            # is if "l" is nested within "i,j"-loops.
+            self.tree = _add_inner_loops(self.tree, outer_loop, not_seen_inames)
+
+        # {{{ update iname to node id
+
+        for iname in outer_loop:
+            self.iname_to_tree_node_id = self.iname_to_tree_node_id.set(
+                iname, outer_loop
+            )
+
+        if inner_loop is not None:
+            for iname in inner_loop:
+                self.iname_to_tree_node_id = self.iname_to_tree_node_id.set(
+                    iname, inner_loop
+                )
+
+        for iname in not_seen_inames:
+            self.iname_to_tree_node_id = self.iname_to_tree_node_id.set(
+                iname, not_seen_inames
+            )
+
+        # }}}
+
+        # }}}
+
+        assert not (outer_redn_inames & redn_inames)
+        return super().map_reduction(
+            expr, expn_state, outer_redn_inames=(outer_redn_inames | redn_inames)
+        )
+
+
+def _add_reduction_loops_in_partial_loop_nest_tree(
+    kernel: LoopKernel, tree: LoopNestTree
+) -> LoopNestTree:
+    """
+    Returns a partial loop nest tree with the loop nests corresponding to the
+    reduction inames added to *tree*.
+    """
+    from loopy.symbolic import SubstitutionRuleMappingContext
+
+    rule_mapping_context = SubstitutionRuleMappingContext(
+        kernel.substitutions, kernel.get_var_name_generator()
+    )
+    reduction_loop_inserter = ReductionLoopInserter(rule_mapping_context, tree)
+
+    def does_insn_have_reduce(kernel, insn, *args):
+        return bool(insn.reduction_inames())
+
+    reduction_loop_inserter.map_kernel(
+        kernel, within=does_insn_have_reduce, map_tvs=False, map_args=False
+    )
+    return reduction_loop_inserter.tree
+
+
+@memoize_on_first_arg
+def _get_partial_loop_nest_tree_for_fusion(kernel: LoopKernel):
+    from loopy.schedule.tools import get_partial_loop_nest_tree
+
+    tree = get_partial_loop_nest_tree(kernel)
+    tree = _add_reduction_loops_in_partial_loop_nest_tree(kernel, tree)
+    return tree
+
+
+def get_kennedy_unweighted_fusion_candidates(
```

Comment:
An alternate discussion to the above is what it would take to allow potentially-nested candidates.

Reply:
I think it's not sound to consider potentially nested candidates.
```
a[i] = 0
b[j] = 1
c[k] = 2
d[i, j] = 3
```
If `i,j,k` are passed in as fusion candidates. `i` and `j` can never be fused and we don't want to design arbitrary heuristic that chooses between fusing `i,k` or `j,k`, better to rely on the user for that.

Reply:
Agree, let's not do that.